File: Core/src/main/java/com/inverce/mod/core/configuration/Preset.java
package com.inverce.mod.core.configuration;

import android.support.annotation.NonNull;

import com.inverce.mod.core.functional.ISupplier;
import com.inverce.mod.core.utilities.SubBuilder;

import java.util.ArrayList;

public class Preset {
    public static Preset.Builder create() {
        // create inner class for specific instance of parent,
        // inner classes have reference to parent (unless static ^^)
        return new Preset().new Builder();
    }

    protected ArrayList<Record<?>> records;

    protected Preset() {
        records = new ArrayList<>();
    }

    @NonNull
    public Preset apply() {
        for (Record record: records) {
            record.apply();
        }
        return this;
    }

    protected class Record<T> {
        Value<T> preference;
        ISupplier<T> value;

        public Record(Value<T> preference, ISupplier<T> value) {
            this.preference = preference;
            this.value = value;
        }

        void apply() {
            preference.set(value.get());
        }
    }

    public class Builder extends SubBuilder<Preset> {
        Builder() {
            super(Preset.this);
        }

        @NonNull
        public <T, Y extends Value<T>> Builder add(Y preference, T value) {
            return addSupplier(preference, () -> value);
        }

        @NonNull
        public <T, Y extends Value<T>> Builder addSupplier(Y preference, ISupplier<T> value) {
            Preset.this.records.add(new Record<>(preference, value));
            return this;
        }
    }
}

File: Core/src/main/java/com/inverce/mod/core/configuration/Preset.kt
package com.inverce.mod.core.configuration

import com.inverce.mod.core.utilities.SubBuilder
import java.util.*

class Preset protected constructor() {
    protected var records: ArrayList<Record<*>> = ArrayList()

    fun apply(): Preset {
        for (record in records) {
            record.apply()
        }
        return this
    }

    protected inner class Record<T>(internal var preference: Value<T>, internal var value: () -> T) {

        internal fun apply() {
            preference.setValue(value())
        }
    }

    inner class Builder internal constructor() : SubBuilder<Preset>(this@Preset) {

        fun <T, Y : Value<T>> add(preference: Y, value: T): Builder {
            return addSupplier(preference, { value })
        }

        fun <T, Y : Value<T>> addSupplier(preference: Y, value: () -> T): Builder {
            this@Preset.records.add(Record(preference, value))
            return this
        }
    }

    companion object {
        fun create(): Preset.Builder {
            // create inner class for specific instance of parent,
            // inner classes have reference to parent (unless static ^^)
            return Preset().Builder()
        }
    }
}